Sources/RedCat/ObservableStore.swift

```diff
@@ -149,12 +170,34 @@ final class ConcreteStore<Reducer : ErasedReducer> : ObservableStore<Reducer.Sta
         }
         
         enqueuedActions = []
-        
+        return newState
     }
     
     @usableFromInline
     override func acceptsAction<Action : ActionProtocol>(_ action: Action) -> Bool {
         reducer.acceptsAction(action)
     }
-    
+	
+		/// Dispatches ```AppDeinit``` and invalidates the receiver.
+		///
+		/// Use this method when your App is about to terminate to trigger cleanup actions.
+		public final func shutDown() {
+				send(Actions.AppDeinit())
+				hasShutdown = true
+		}
+	
```

Comment:
I've been thinking about adding another method to reducers:

```
func willChangeState<Action : ActionProtocol>(_ action: Action, state: State) -> Bool
```

The default implementation would just return true. However, I discarded this idea because for reducers that actually only sometimes react to their action, you usually gather more and more data using ```guard let``` and ```guard case```, so ultimately, users may be inclined to just not implement the ```willChangeState``` method even though this technically makes their reducer incorrect.

Revisiting the issue, I thought of the following: most reducers have an associated ```Action``` type; for all those reducers, one could add another ```UnguardedAction``` type that defaults to ```Action``` and a method

```
func extractChanges(from action: UnguardedAction, for state: State) -> Action?
```

Those Reducers can then provide a default implementation for ```willChangeState```:

```
func willChangeState<Action: ActionProtocol>(_ action: Action, state: State) -> Bool {
   (action as? UnguardedAction).flatMap{extractChanges(from: $0, for: state) != nil
}
```

It'll add some complexity for ```DispatchReducer```s though.

The ultimate benefit of all this would be that we can prevent a notification cycle if the action doesn't impact the state. Moreover, for all reducers that don't implement the method, this calculation can be done at compile time, which makes this a rather cheap feature. The only downer is that ```extractChanges``` needs to be executed twice per action (for those reducers where this can't e optimized out). Suggestions would be highly appreciated.

Reply:
hmm since the subscriber can observe a part of a state (there may be situations when the state has changed but a substate has not), it is likely that the subscriber should handle those situations when updates is expensive: skip equal substates, etc

Reply:
Yep.
Highly unusual:
```
statePublisher.removeDuplicates()
```
but
```
statePublisher.map(\.partialState).removeDuplicates()
```
is quite idiomatic. One could add the above described sanity checks though (if the double call issue is addressed).